### DISCUSSION TITLE
Persist using initial state

### DISCUSSION THREAD

**Discussion by HermanPierre** [pmndrs/zustand#2619] [Q&A] [answered]
Hello, 
I am currently using NextJS and zustand with the persist middleware:
**Store.tsx:**
```
import {create} from 'zustand';
import {persist} from "zustand/middleware";

// Définir le type de l'état
interface State {
    user: User | null;
    isLoggedIn: boolean;
    setUser: (user: User) => void;
}

// Créer le store avec Zustand
const useUserStore = create<State>()(persist((set) => ({
    isLoggedIn: false,
    user: null,
    setUser: (user: User) => set((state: State) => ({ isLoggedIn: true, user: user })),
}), {
    name: 'user-storage',
    // getStorage: () => localStorage
}));


export default useUserStore;

```
**Layout.tsx**
```
'use client'

import 'bootstrap/dist/css/bootstrap.min.css';
import "./globals.css";
import {Toaster} from "react-hot-toast";
import React, {useEffect} from "react";
import useUserStore from "@/store/userStore";
import {usePathname, useRouter} from "next/navigation";

export default function RootLayout({children,}: Readonly<{
    children: React.ReactNode;
}>) {

    const routerPath = usePathname();
    const router = useRouter();
    const isLoggedIn = useUserStore(state => state.isLoggedIn);

    useEffect(() => {
        console.log({isLoggedIn})
        if (!isLoggedIn && routerPath !== '/login') {
            router.push('/login');
        }
    }, [isLoggedIn, routerPath]);

    return (
        <>
            <html lang="fr">
            <head>
            </head>
            <body>
            {children}
            <Toaster/>
            </body>
            </html>
        </>
    );
}

```

The problem here is that when I use the deprecated "getStorage" with localStorage, it works well, when I refresh my store gets immediatly populated with the localStorage values if it exists.
But if I use "storage" which is now enabled by default, my useEffect in my layout first get the default version of my store then it gets the store populated with the localStorage values, but to check if my user is logged in I need to have the boolean at true at the first version of my store... 

<img width="605" alt="image" src="https://github.com/pmndrs/zustand/assets/32395678/9f10e652-991f-4080-ae0c-4d435b588686">

How should I make it work ? How to avoid the store using the default store before being sure that localStorage isn't populated ?
Thanks :)

**Comment by qiushiyan** [comment]
I am having the same problem. I am following the pattern in the [docs](https://docs.pmnd.rs/zustand/guides/nextjs) to create a store with initialization parameters and provide it through a context provider.

For anyone interested in a minimal example, see [here](https://codesandbox.io/p/github/qiushiyan/next-zustand-persist/draft/hungry-gould?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxvf77e500073b6jxwjxmgnl%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxvf77e400023b6jjjosrx37%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxvf77e400053b6jq3nprdc2%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxvf77e500063b6jw7xoh58t%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxvf77e400023b6jjjosrx37%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxvf77e400013b6j8yevy098%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clxvf77e400023b6jjjosrx37%2522%252C%2522activeTabId%2522%253A%2522clxvf77e400013b6j8yevy098%2522%257D%252C%2522clxvf77e500063b6jw7xoh58t%2522%253A%257B%2522id%2522%253A%2522clxvf77e500063b6jw7xoh58t%2522%252C%2522activeTabId%2522%253A%2522clxvf7a7m002b3b6jpulg1n9t%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clxvf7a7m002b3b6jpulg1n9t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clxvf77e400053b6jq3nprdc2%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxvf77e400033b6jaogzj96s%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clxvf77e400043b6j5l5ogqfg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clxvf2pnl000vdbij8ac15fp3%2522%257D%255D%252C%2522id%2522%253A%2522clxvf77e400053b6jq3nprdc2%2522%252C%2522activeTabId%2522%253A%2522clxvf77e400033b6jaogzj96s%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D). Some details:

- I am generating dynamic routes and create a counter store for each of them, `/1` will have initial count of 1,  `/2` the initial count of 2, and so on. 

```tsx
// [id]/page.tsx

import { Counter } from "@/components/counter";
import { CounterStoreProvider } from "@/components/provider";

type Props = {
	params: {
		id: string;
	};
};

export default function ({ params }: Props) {
	return (
		<CounterStoreProvider count={Number(params.id)}>
			<Counter />
		</CounterStoreProvider>
	);
}
```

The persistence is keyed by the initial count

```ts
// store/counter.ts

export const createCounterStore = (initState: CounterState) => {
	return createStore<CounterStore>()(
		persist(
			(set) => ({
				count: initState.count,
				decrementCount: () => set((state) => ({ count: state.count - 1 })),
				incrementCount: () => set((state) => ({ count: state.count + 1 })),
			}),
			{
				name: `count-page-${initState.count}`,
				storage: createJSONStorage(() => localStorage),
			},
		),
	);
};
```

The provider component is set up the same way in the docs. 

When you increase the counter and then refresh the page, the counter starts as the initial value and then quickly jumps to the stored value.

https://github.com/pmndrs/zustand/assets/48123899/d801aeb1-6161-4ac9-bca2-5f8f57b96e8f

I tried manually rehydrate the store, but it does not work. 

It also seems that this only applies to dev server, in prod the counter starts straight from the storage value.

**Reply by dbritto-dev** [reply]
@qiushiyan here you go: https://stackblitz.com/edit/nextjs-and-zustand-persist?file=README.md

> Note: btw, the docs example is not ready to use with persist middleware.

**Reply by HermanPierre** [reply]
@dbritto-dev why having to handle hydration manually if we didn't had to before ? (with getStorage I mean)

**Reply by dbritto-dev** [reply, marked as answer]
@HermanPierre that's wrong, you should always had to handle that. There's no localStorage on server and you need to make sure that server and client first render are the same. On the client after first render you could hydrate and get the current state from the browser. 

> Note: remember that you need to render twice one on the server and one on the client when you use server mechanism/solution like SSR, SSG, etc

**Reply by HermanPierre** [reply]
@dbritto-dev Oh okay but I never wanted to have the store on the server side, I want it on the client side. It seems way harder to handle user experience data on the server side than on the client side

**Reply by dbritto-dev** [reply]
@HermanPierre yes and no, you need to know the trade-offs

**Reply by HermanPierre** [reply]
@dbritto-dev okay I've watched a lot of videos on NextJS last few hours I understand better, thanks :)

**Reply by dbritto-dev** [reply]
Just ping me if you have any questions

**Reply by qiushiyan** [reply]
@dbritto-dev Thanks for the reply!

**Comment by Debbl** [comment]
I'm having the same problem. I cannot get the initial value immediately when the page is refreshed. 

In this code, I have to set the editor's value and change the value when the editor updates. but I cannot get correct value for editor.

<img width="802" alt="image" src="https://github.com/pmndrs/zustand/assets/61053131/44b008a2-377c-4a38-bb1c-527e9e4bf256">

Here's the same case in a different version.
## 4.4.7 version
- https://codesandbox.io/p/live/de15e471-b627-4af0-82e0-2e2d8dd08d05
- click setValue
- refresh page
- value is `aaaaa`

## 4.5.0 version
- https://codesandbox.io/p/live/760f9114-2014-4504-9817-10c8bab38f85
- click setValue
- refresh page
- value is ''

**Reply by dbritto-dev** [reply]
@Debbl you mind creating them on stackblitz? I wasn't able to open them

**Reply by Debbl** [reply]
## 4.4.7
- https://stackblitz.com/edit/stackblitz-starters-7h3yzy?file=app%2Fpage.tsx

## 4.5.0
- https://stackblitz.com/edit/stackblitz-starters-rjqkzf?file=app%2Fpage.tsx

**Reply by Debbl** [reply]
I tried to write it using this and it is working.

<img width="728" alt="image" src="https://github.com/pmndrs/zustand/assets/61053131/b7666c4d-30a6-4a00-aae6-4e1d00b06594">

**Reply by dbritto-dev** [reply]
@Debbl btw, did you follow our nextjs guide? https://docs.pmnd.rs/zustand/guides/nextjs

**Reply by dbritto-dev** [reply]
well, I saw that your are not following our recommendations:

- avoid global stores
- use Context API to isolate/create store per request
- skip hydration and then hydrate on client

**Reply by dbritto-dev** [reply]
Here you go: https://stackblitz.com/edit/nextjs-and-zustand-persist?file=README.md

**Reply by Debbl** [reply]
However, I still can't get the correct value in the useEffect. Is this expected behavior?
https://stackblitz.com/edit/nextjs-and-zustand-persist-7llgcr?file=app%2F_components%2Fhome-page.tsx

<img width="676" alt="image" src="https://github.com/pmndrs/zustand/assets/61053131/34fcd102-1add-4a96-b0a9-365ae6ea71ef">

**Reply by HermanPierre** [reply]
> well, I saw that your are not following our recommendations:
> 
> * avoid global stores
> * use Context API to isolate/create store per request
> * skip hydration and then hydrate on client

I don't understand the purpose of creating a context to use a Store ? Having a store is by definition a way to avoid being dependent of a context to use some informations somewhere in the app ain't it ?

**Reply by dbritto-dev** [reply]
@HermanPierre we mentioned that in our nextjs guide, lmk if something isn't clear

**Reply by dbritto-dev** [reply]
but in a nutshell you should avoid next things:
- global stores (sharing data across requests)
- read/write stores on server
- stuff that could change on client and server like: document or window

**Reply by HermanPierre** [reply]
> but in a nutshell you should avoid next things:
> 
> * global stores (sharing data across requests)
> * read/write stores on server
> * stuff that could change on client and server like: document or window

Okay thanks, the last two are easy to avoid but for the first one i'm trying as much as I can to avoid having global stores in my app but it ain't always easy ahah
Thanks :)

**Reply by dbritto-dev** [reply]
@HermanPierre yeah, is not easy. Actually, we are working on updating docs content and adding some learn docs stuff to talk about ways of dealing with that

**Reply by tshmieldev** [reply]
sooo
Any way I can get the initial value from localStorage so the first read is not the default?
I really want the first read to already have the data... IG I could make a wrapper to just
if(state.loading) return localStorage.get*...*
return state

but this seems super super scuffed :( I just want to read state and have the data immediately
